### PR TITLE
[WIP]Add packer support for building AWS AMI

### PR DESCRIPTION
**Pull request by shub-kris**
This PR aims to fix issue #433.

### CODE REVIEW

File: .github/workflows/build-ami.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+        - uses: actions/checkout@v3
+        - uses: hashicorp/setup-packer@main
+        
+        - name: configure aws credentials
+          uses: aws-actions/configure-aws-credentials@v1
+          with:
+            aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+            aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+            aws-region: us-east-1
```
Comment (philschmid) [inline .github/workflows/build-ami.yml:30]: @glegendre01 Is that okay with you? The CI runs a packer inside GitHub actions to create an AMI on AWS. Or do we use [Assume Role directly using GitHub OIDC provider](https://github.com/aws-actions/configure-aws-credentials?tab=readme-ov-file#using-this-action)

File: .github/workflows/build-ami.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+name: Build AWS Neuron AMI
+
+on:
+    push:
+      branches:
+        - main
+      paths:
+        - 'aws-ami-creation/**'
+  workflow_dispatch:
```
Comment (philschmid) [inline .github/workflows/build-ami.yml:9]: We should be able to pass which "tag" to build. That on a cron schedule it builds the main branch but that we can use the same action for building a release image we can then publish.

File: aws-ami-creation/README.md
Comment (philschmid) [inline aws-ami-creation/README.md:1]: What about moving it into an `infrastructure/` folder with a dedicated `ami/` folder? We might have additional infrastructure components in the future.

File: aws-ami-creation/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+## Folder Structure
+
+- `hcl2-files/`: This directory contains various files related to packer.
+- `scripts/`: This directory contains scripts used by packer provisner for installing additonal packages/softwares.
+
+
+## Usage
+
+[Add instructions or usage guidelines here if applicable]
```
Comment (philschmid) [inline aws-ami-creation/README.md:9]: We should how to build with cli and through the CI so other team members can build the AMIs too.

File: aws-ami-creation/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+## Folder Structure
+
+- `hcl2-files/`: This directory contains various files related to packer.
+- `scripts/`: This directory contains scripts used by packer provisner for installing additonal packages/softwares.
```
Comment (philschmid) [inline aws-ami-creation/README.md:4]: can we make this a bit more specific if you are not familiar with packer?

File: aws-ami-creation/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+## Folder Structure
```
Comment (philschmid) [inline aws-ami-creation/README.md:1]: Also lets add some "header" -> Content for build AMI with ....

File: aws-ami-creation/hcl2-files/sources.pkr.hcl [hunk 6]
```diff
@@ -0,0 +1,14 @@
+source "amazon-ebs" "ubuntu" {
+  ami_name      = "huggingface-neuron-dl-ami-ubuntu-22.04_{{timestamp}}"
+  instance_type = var.instance_type
+  region        = var.region
+  source_ami    = var.source_ami
+  ssh_username  = var.ssh_username
+  launch_block_device_mappings {
+    device_name           = "/dev/sda1"
+    encrypted             = true
+    volume_size           = 100
```
Comment (philschmid) [inline aws-ami-creation/hcl2-files/sources.pkr.hcl:10]: is that big enough?

File: aws-ami-creation/hcl2-files/variables.pkr.hcl [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+variable "instance_type" {
+  default     = "trn1.2xlarge"
+  description = "EC2 machine type for building AMI"
+  type        = string
+}
+
+variable "source_ami" {
+  default     = "ami-0328507805a665ec1"
```
Comment (philschmid) [inline aws-ami-creation/hcl2-files/variables.pkr.hcl:14]: can we derive this automatically to use the latest released AWS image? then we don't need to check, e.g. with the AWS CLI you can do 
```bash
aws ec2 describe-images --region us-east-1 --owners amazon --filters 'Name=name,Values=Deep Learning AMI Neuron PyTorch 1.13 (Ubuntu 20.04) ????????' 'Name=state,Values=available' --query 'reverse(sort_by(Images, &CreationDate))[:1].ImageId' --output text
```

File: aws-ami-creation/scripts/install_huggingface_libraries.sh [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+
+echo "Step: copy-optimum-neuron-examples"
+git clone -b "v0.0.17" https://github.com/huggingface/optimum-neuron.git
+mkdir /home/ubuntu/huggingface-neuron-samples/
+mv optimum-neuron/examples/* /home/ubuntu/huggingface-neuron-samples/
+rm -rf optimum-neuron
+chown -R ubuntu:ubuntu /home/ubuntu/huggingface-neuron-samples
+
+
```
Comment (philschmid) [inline aws-ami-creation/scripts/install_huggingface_libraries.sh:31]: we need to add the "notebooks"

File: aws-ami-creation/scripts/install_huggingface_libraries.sh [hunk 9]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+echo "Step: install-hugging-face-libraries"
+
+echo "Activating the virtual-env"
+source /opt/aws_neuron_venv_pytorch/bin/activate
```
Comment (philschmid) [inline aws-ami-creation/scripts/install_huggingface_libraries.sh:6]: This would be a big breaking change to the existing AMI. Where you can straight up use "python". Can we either activate it by default or not use a virtual env?

**Review by philschmid — COMMENTED** [review]
Left some comments! 

Do you know if we also can add a "print" when the instance is launched with some instructions, e.g. where to find examples, docs and how to run something. Like a welcome messages.

File: aws-ami-creation/hcl2-files/build.pkr.hcl [hunk 10]
```diff
@@ -0,0 +1,17 @@
+build {
+  name = "build-hf-dl-neuron"
+  sources = [
+    "source.amazon-ebs.ubuntu"
+  ]
+  // provisioner "shell" {
+  //   inline = [
+  //     "echo foo"
+  //   ]
+  // }
```
Comment (dacorvo) [inline aws-ami-creation/hcl2-files/build.pkr.hcl]: I assume this is a leftover of previous attempts. If so consider removing it.

File: aws-ami-creation/hcl2-files/variables.pkr.hcl (hunk 7 above)
Comment (dacorvo) [reply aws-ami-creation/hcl2-files/variables.pkr.hcl:14]: I think we should keep the ability to regenerate previous AMIs, so it is IMHO better to have the exact version of the source AMI we used somewhere.
In other words, if we can put that in another file that calls this, then I am fine with using another default. Otherwise I would keep it.

File: aws-ami-creation/hcl2-files/sources.pkr.hcl (hunk 6 above)
Comment (shub-kris) [reply aws-ami-creation/hcl2-files/sources.pkr.hcl:10]: Yes, I tried and tested it.

File: aws-ami-creation/hcl2-files/build.pkr.hcl (hunk 10 above)
Comment (shub-kris) [reply aws-ami-creation/hcl2-files/build.pkr.hcl]: It's already removed :)

File: aws-ami-creation/scripts/install_huggingface_libraries.sh (hunk 9 above)
Comment (shub-kris) [reply aws-ami-creation/scripts/install_huggingface_libraries.sh:6]: The base AMI let's us use the neuron package and other pre-installed packages only if we activate this environment. If we plan to not use a virtual env then we need to install everything just like the earlier repo.

File: .github/workflows/build-ami.yml (hunk 1 above)
Comment (glegendre01) [reply .github/workflows/build-ami.yml:30]: first good point, the runner is not self-hosted, so it's ephemeral
Second point, if your ACCESS Key authorization are scoped to packer only it's fine, if it's not the case, we need to do it.
Indeed, the good solution is too use OIDC, i added on our todo list

File: aws-ami-creation/scripts/install_huggingface_libraries.sh (hunk 9 above)
Comment (shub-kris) [reply aws-ami-creation/scripts/install_huggingface_libraries.sh:6]: And in that case we should use https://aws.amazon.com/releasenotes/aws-deep-learning-ami-base-neuron-ubuntu-20-04/ as mentioned in the documentation that 

```
The Base AMI comes with a foundational platform of GPU drivers and acceleration libraries to deploy your own customized deep learning environment. 
```

File: aws-ami-creation/hcl2-files/variables.pkr.hcl (hunk 7 above)
Comment (shub-kris) [reply aws-ami-creation/hcl2-files/variables.pkr.hcl:14]: I would also agree to keep the exact version, as it will help in debugging if we ran into any kind of problem in future.

File: .github/workflows/build-ami.yml (hunk 1 above)
Comment (philschmid) [reply .github/workflows/build-ami.yml:30]: So its okay for now to continue until we have OIDC?

File: aws-ami-creation/hcl2-files/variables.pkr.hcl (hunk 7 above)
Comment (philschmid) [reply aws-ami-creation/hcl2-files/variables.pkr.hcl:14]: SGTM! But then lets include it in the "github" actions as parameter that you could overwrite it if needed.

File: .github/workflows/build-ami.yml (hunk 1 above)
Comment (glegendre01) [reply .github/workflows/build-ami.yml:30]: yes

**Comment by shub-kris** [comment]
> Left some comments!
> 
> Do you know if we also can add a "print" when the instance is launched with some instructions, e.g. where to find examples, docs and how to run something. Like a welcome messages.

Hi @philschmid regarding this, I tried making a file called 99-custom-message in `/etc/update-motd.d`, it doesn't work when ssshing for the first time but works in the second time and I have no clue why? 

I also tried doing something like this 

```
  provisioner "shell" {
    inline = [
      "sudo mv /tmp/welcome.sh /etc/ssh/welcome.sh",
      "sudo chmod +x /etc/ssh/welcome.sh",
      "sudo sed -i 's/#Banner/Banner/' /etc/ssh/sshd_config",
      "sudo sed -i '/^Banner/c\\Banner /etc/ssh/welcome.sh' /etc/ssh/sshd_config",
      "sudo systemctl restart sshd"
    ]
  }
```
This works but the script is not seen as a script but it rather prints out everything. 

So if my `/tmp/welcome.sh` was like this 

```
#!/bin/bash

"Welcome to your instance!"
"You can find examples and documentation at <link>"
"To run something, use <command>"

```
Then everything is printed as it is

**Comment by philschmid** [comment]
> > Left some comments!
> > Do you know if we also can add a "print" when the instance is launched with some instructions, e.g. where to find examples, docs and how to run something. Like a welcome messages.
> 
> Hi @philschmid regarding this, I tried making a file called 99-custom-message in `/etc/update-motd.d`, it doesn't work when ssshing for the first time but works in the second time and I have no clue why?
> 
> I also tried doing something like this
> 
> ```
>   provisioner "shell" {
>     inline = [
>       "sudo mv /tmp/welcome.sh /etc/ssh/welcome.sh",
>       "sudo chmod +x /etc/ssh/welcome.sh",
>       "sudo sed -i 's/#Banner/Banner/' /etc/ssh/sshd_config",
>       "sudo sed -i '/^Banner/c\\Banner /etc/ssh/welcome.sh' /etc/ssh/sshd_config",
>       "sudo systemctl restart sshd"
>     ]
>   }
> ```
> 
> This works but the script is not seen as a script but it rather prints out everything.
> 
> So if my `/tmp/welcome.sh` was like this
> 
> ```
> #!/bin/bash
> 
> "Welcome to your instance!"
> "You can find examples and documentation at <link>"
> "To run something, use <command>"
> ```
> 
> Then everything is printed as it is

Please try look into what it is not shown. Having a message for the second ssh is not helpful. Have you tried searching for it?, e.g. https://askubuntu.com/questions/1394600/motd-not-showing-up-on-ubuntu-21-10
Or does the base AMI already have something which can lead to issues etc. It should work since i used other ubuntu based ami in the past which had this.

File: .github/workflows/build-ami.yml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    inputs: 
+      tag: 
+        description: 'Tag to use for the AMI'
```
Comment (philschmid) [inline .github/workflows/build-ami.yml:12]: should have a default value to `main`

File: .github/workflows/build-ami.yml [hunk 12]
```diff
@@ -0,0 +1,62 @@
+name: Build AWS Neuron AMI
+
+on:
+    push:
+      branches:
+        - main
+      paths:
+        - 'aws-ami-creation/**'
```
Comment (philschmid) [inline .github/workflows/build-ami.yml:8]: We should add "pull_request" too otherwise you cannot validate if the changes work before merging.

File: .github/workflows/build-ami.yml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+jobs:
+    build-ami:
+        defaults:
+          run:
+            working-directory: infrastructure/ami
+        runs-on: ubuntu-latest
+        steps:
+
+        - uses: actions/checkout@v3
```
Comment (philschmid) [inline .github/workflows/build-ami.yml:25]: this should checkout based on the version you passed in.

File: infrastructure/ami/README.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+You can also install Packer for other OS from [here](https://developer.hashicorp.com/packer/tutorials/docker-get-started/get-started-install-cli).
+
+#### Configure AWS Credentials
+
+Using Environment Variables:
+```bash
+export AWS_ACCESS_KEY_ID=<access_key>
+export AWS_SECRET_ACCESS_KEY=<secret_key>
+````
```
Comment (philschmid) [inline infrastructure/ami/README.md:37]: There is ` to much

File: infrastructure/ami/scripts/setup-neuron.sh [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    software-properties-common \
+    unzip \
+    vim \
+    wget \
+    zlib1g-dev \
+    libsndfile1-dev \
+    ffmpeg \
+    g++
+
+sudo apt-get update # Update to fix broken packages
```
Comment (philschmid) [inline infrastructure/ami/scripts/setup-neuron.sh:26]: Why is this needed? Aren't we doing this 10 lines above?

File: infrastructure/ami/scripts/setup-neuron.sh [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    --extra-index-url https://pip.repos.neuron.amazonaws.com \
+    "protobuf==3.20.2" \
+    torch-neuronx=="1.13.1.1.13.0" \
+    "transformers-neuronx==0.9.474" \
+    "neuronx_distributed==0.6.0" \
+    "tensorboard-plugin-neuronx" \
+    "torchvision==0.14.*"'
+
+sudo -H -u ubuntu bash -c 'echo 'export PATH="${HOME}/.local/bin:$PATH"' >> ${HOME}/.bashrc'
+sudo -H -u ubuntu bash -c 'python -c "import torch_neuronx"'
```
Comment (philschmid) [inline infrastructure/ami/scripts/setup-neuron.sh:40]: same as above why sudo + user?

**Review by philschmid — CHANGES_REQUESTED** [review]
I did another review. Please look at my comments there are a lot of things unclear to me. Also we want to use the existing Neuron AMI. We don't want to install any neuronx dependencies.

File: aws-ami-creation/scripts/install_huggingface_libraries.sh (hunk 9 above)
Comment (philschmid) [reply aws-ami-creation/scripts/install_huggingface_libraries.sh]: If those are only installed in `venv` then we should defaulty activate those on instance start and ssh connection.

File: infrastructure/ami/hcl2-files/variables.pkr.hcl [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+variable "instance_type" {
+  default     = "trn1.2xlarge"
+  description = "EC2 machine type for building AMI"
+  type        = string
+}
+
+variable "source_ami" {
+  default     = "ami-077399be2d0ae7e0c"
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/hcl2-files/variables.pkr.hcl:14]: If we assign something as a variable in packer, we have to provide a default value else it throws an error while validating

```
Error: Unset variable "source_ami"

A used variable must be set or have a default value; see
```

[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/hcl2-files/variables.pkr.hcl:14]: Okay, I am setting all the default values in variables.pkr.hcl

File: infrastructure/ami/scripts/install-hugginface-libraries.sh [hunk 18]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+echo "Step: install-hugging-face-libraries"
+
+sudo -H -u ubuntu bash -c 'pip install --upgrade --no-cache-dir \
+    "transformers[sklearn,sentencepiece,vision]==4.36.2" \
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/install-hugginface-libraries.sh:6]: @philschmid do you mean the packages and the versions both or just the versions?

[the remark this answers is not in the document]
Comment (philschmid) [reply infrastructure/ami/scripts/install-hugginface-libraries.sh:6]: i would go with transformers and optimum for now

File: infrastructure/ami/scripts/setup-neuron.sh [hunk 19]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/setup-neuron.sh:1]: Yes

File: infrastructure/ami/scripts/setup-neuron.sh [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+echo "Step: validate-neuron-devices"
+neuron-ls
+
+echo "Step: install-ubuntu-packages"
+sudo apt-get update
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/setup-neuron.sh:6]: Removing neuron packages installation

File: infrastructure/ami/scripts/setup-neuron.sh (hunk 15 above)
Comment (shub-kris) [reply infrastructure/ami/scripts/setup-neuron.sh:26]: It will be removed

File: infrastructure/ami/scripts/setup-neuron.sh [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    unzip \
+    vim \
+    wget \
+    zlib1g-dev \
+    libsndfile1-dev \
+    ffmpeg \
+    g++
+
+sudo apt-get update # Update to fix broken packages
+sudo apt-get install -y --no-install-recommends python-is-python3 python3-pip python3-dev python3-virtualenv
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/setup-neuron.sh:27]: Will be removed

File: infrastructure/ami/scripts/setup-neuron.sh [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+
+echo "Step: install-neuron-python-packages"
+sudo -H -u ubuntu bash -c 'pip install --upgrade --no-cache-dir \
+    --extra-index-url https://pip.repos.neuron.amazonaws.com \
+    "protobuf==3.20.2" \
+    torch-neuronx=="1.13.1.1.13.0" \
+    "transformers-neuronx==0.9.474" \
+    "neuronx_distributed==0.6.0" \
+    "tensorboard-plugin-neuronx" \
+    "torchvision==0.14.*"'
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/setup-neuron.sh:37]: No more needed

File: infrastructure/ami/scripts/setup-neuron.sh (hunk 16 above)
Comment (shub-kris) [reply infrastructure/ami/scripts/setup-neuron.sh:40]: No more needed

File: infrastructure/ami/scripts/install-hugginface-libraries.sh [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+rm -rf optimum-neuron
+chown -R ubuntu:ubuntu /home/ubuntu/huggingface-neuron-samples /home/ubuntu/huggingface-neuron-notebooks
+
+
+echo "Step: clean-apt-cache"
+sudo apt autoremove -y
+sudo apt clean -y
+
+echo "Step: validate-imports-of-huggingface-libraries"
+sudo -H -u ubuntu bash -c 'python -c "import transformers;import datasets;import accelerate;import evaluate;import tensorboard; import torch;"'
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/install-hugginface-libraries.sh:35]: No more needed

File: infrastructure/ami/scripts/install-hugginface-libraries.sh [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+mkdir /home/ubuntu/huggingface-neuron-samples/ /home/ubuntu/huggingface-neuron-notebooks/
+mv optimum-neuron/examples/* /home/ubuntu/huggingface-neuron-samples/
+mv optimum-neuron/notebooks/* /home/ubuntu/huggingface-neuron-notebooks/
+rm -rf optimum-neuron
+chown -R ubuntu:ubuntu /home/ubuntu/huggingface-neuron-samples /home/ubuntu/huggingface-neuron-notebooks
+
+
+echo "Step: clean-apt-cache"
+sudo apt autoremove -y
+sudo apt clean -y
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/install-hugginface-libraries.sh:32]: No more needed

File: infrastructure/ami/scripts/install-hugginface-libraries.sh [hunk 25]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+echo "Step: install-hugging-face-libraries"
+
+sudo -H -u ubuntu bash -c 'pip install --upgrade --no-cache-dir \
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/install-hugginface-libraries.sh:5]: No more needed, yes packer uses `ubuntu` user

File: .github/workflows/build-ami.yml [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+        
+        - name: Packer Validate
+          id: validate
+          run: packer validate hcl2-files
+          continue-on-error: true
+        
+        - name: Packer Build
+          id: build
+          run:  | 
+            SOURCE_AMI=ami-077399be2d0ae7e0c
```
[the remark this answers is not in the document]
Comment (shub-kris) [reply .github/workflows/build-ami.yml:60]: Putting only inside variables.pkr.hcl

File: .github/workflows/build-ami.yml [hunk 27]
```diff
@@ -61,4 +61,14 @@ jobs:
...
               packer build -var "optimum_neuron_tag=${{ steps.determine-tag.outputs.TAG }}" hcl2-files
-        
\ No newline at end of file
+
+          - name: Slack Notification on Failure
+          if: ${{ failure() && github.event_name == 'schedule' }}
+          uses: rtCamp/action-slack-notify@v2
+          env:
+            SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
+            SLACK_CHANNEL: ${{ secrets.SLACK_CHANNEL }}
```
[the remark this answers is not in the document]
Comment (philschmid) [reply .github/workflows/build-ami.yml:70]: can you check with the infra team. They have some convetions for alerts and they can create the channel

[the remark this answers is not in the document]
Comment (shub-kris) [reply .github/workflows/build-ami.yml:70]: Okay

File: infrastructure/ami/scripts/welcome-msg.sh
[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/welcome-msg.sh:1]: Yes

[the remark this answers is not in the document]
Comment (shub-kris) [reply infrastructure/ami/scripts/welcome-msg.sh:1]: It prints this 

=============================================================================
       __|  __|_  )
       _|  (     /   HuggingFace Deep Learning Neuron AMI (Ubuntu 20.04)
      ___|\___|___|
=============================================================================
Welcome to the HuggingFace Deep Learning Neuron AMI (Ubuntu 20.04)
* Examples:     /home/ubuntu/huggingface-neuron-samples 
* Notebooks:    /home/ubuntu/huggingface-neuron-notebooks 
* Documentation: https://huggingface.co/docs/optimum-neuron/ 
=============================================================================

[the remark this answers is not in the document]
Comment (philschmid) [reply infrastructure/ami/scripts/welcome-msg.sh:1]: nice

**Review by philschmid — APPROVED** [review]
LGTM!

File: .github/workflows/build-ami.yml (hunk 1 above)
Comment (shub-kris) [reply .github/workflows/build-ami.yml]: @philschmid @glegendre01 how can we add this secrets then?

Comment (philschmid) [reply .github/workflows/build-ami.yml]: @shub-kris, what scope (iam permissions) do you need? @glegendre01 should be able to add those

Comment (shub-kris) [reply .github/workflows/build-ami.yml]: The ones mentioned in the offical doc here: https://developer.hashicorp.com/packer/integrations/hashicorp/amazon#iam-task-or-instance-role

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "ec2:AttachVolume",
        "ec2:AuthorizeSecurityGroupIngress",
        "ec2:CopyImage",
        "ec2:CreateImage",
        "ec2:CreateKeyPair",
        "ec2:CreateSecurityGroup",
        "ec2:CreateSnapshot",
        "ec2:CreateTags",
        "ec2:CreateVolume",
        "ec2:DeleteKeyPair",
        "ec2:DeleteSecurityGroup",
        "ec2:DeleteSnapshot",
        "ec2:DeleteVolume",
        "ec2:DeregisterImage",
        "ec2:DescribeImageAttribute",
        "ec2:DescribeImages",
        "ec2:DescribeInstances",
        "ec2:DescribeInstanceStatus",
        "ec2:DescribeRegions",
        "ec2:DescribeSecurityGroups",
        "ec2:DescribeSnapshots",
        "ec2:DescribeSubnets",
        "ec2:DescribeTags",
        "ec2:DescribeVolumes",
        "ec2:DetachVolume",
        "ec2:GetPasswordData",
        "ec2:ModifyImageAttribute",
        "ec2:ModifyInstanceAttribute",
        "ec2:ModifySnapshotAttribute",
        "ec2:RegisterImage",
        "ec2:RunInstances",
        "ec2:StopInstances",
        "ec2:TerminateInstances"
      ],
      "Resource": "*"
    }
  ]
}


```

File: infrastructure/ami/README.md [hunk 28]
```diff
@@ -0,0 +1,78 @@
+# Building AMI with Packer
+
+This directory contains the files for building AMI using [Packer](https://github.com/hashicorp/packer) that is later published as a AWS Marketplace asset.
+
+
+## Folder Structure
+
+- [hcl2-files](./hcl2-files/) - Includes different files which are used by a Packer pipeline to build an AMI. The files are:
+  - [build.pkr.hcl](./hcl2-files/build.pkr.hcl): contains the [build](https://developer.hashicorp.com/packer/docs/templates/hcl_templates/blocks/build) defines what builders are started, how to provision them using [provisioner](https://developer.hashicorp.com/packer/docs/templates/hcl_templates/blocks/build/provisioner) and if necessary what to do with their artifacts using `post-process`.
```
Comment (dacorvo) [inline infrastructure/ami/README.md:9]:
```suggestion
  - [build.pkr.hcl](./hcl2-files/build.pkr.hcl): contains the [build](https://developer.hashicorp.com/packer/docs/templates/hcl_templates/blocks/build), defines what builders are started, how to provision them using [provisioner](https://developer.hashicorp.com/packer/docs/templates/hcl_templates/blocks/build/provisioner) and if necessary what to do with their artifacts using `post-process`.
```

File: infrastructure/ami/hcl2-files/variables.pkr.hcl [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+variable "instance_type" {
+  default     = "trn1.2xlarge"
+  description = "EC2 machine type for building AMI"
+  type        = string
+}
+
+variable "source_ami" {
+  default     = "ami-0fbea04d7389bcd4e" # To get latest AMI use: aws ec2 describe-images --region us-east-1 --owners amazon --filters 'Name=name,Values=Deep Learning AMI Neuron PyTorch 1.13 (Ubuntu 20.04) ????????' 'Name=state,Values=available' --query 'reverse(sort_by(Images, &CreationDate))[:1].ImageId' --output text
```
Comment (dacorvo) [inline infrastructure/ami/hcl2-files/variables.pkr.hcl:13]: Consider splitting this comment on multiple lines: not all editors wrap lines automatically (mine doesn't)

File: infrastructure/ami/hcl2-files/variables.pkr.hcl [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+
+variable "transformers_version" {
+  default     = "4.36.2"
+  description = "Transformers version to install"
+  type        = string
+}
+
+variable "ami_users" {
+  default     = ["754289655784", "558105141721"]
+  description = "AWS accounts to share AMI with"
```
Comment (dacorvo) [inline infrastructure/ami/hcl2-files/variables.pkr.hcl:37]: Are we sure we want to share this on a public repository ?

File: infrastructure/ami/scripts/install-huggingface-libraries.sh [hunk 31]
```diff
@@ -0,0 +1,39 @@
...
+# Activate the neuron virtual environment
+source /opt/aws_neuron_venv_pytorch/bin/activate
+
+echo "Step: install-hugging-face-libraries"
+
+echo "TRANSFORMERS_VERSION: $TRANSFORMERS_VERSION"
+echo "OPTIMUM_VERSION: $OPTIMUM_VERSION"
+
+pip install --upgrade --no-cache-dir \
+    "transformers[sklearn,sentencepiece,vision]==$TRANSFORMERS_VERSION" \
```
Comment (dacorvo) [inline infrastructure/ami/scripts/install-huggingface-libraries.sh:12]: Why not letting optimum-neuron pull the right version ?

File: infrastructure/ami/hcl2-files/variables.pkr.hcl (hunk 30 above)
Comment (shub-kris) [reply infrastructure/ami/hcl2-files/variables.pkr.hcl:37]: Umm, then we add it in secrets somewhere I guess. 
@philschmid wdyt?

Comment (philschmid) [reply infrastructure/ami/hcl2-files/variables.pkr.hcl:37]: Yes AWS Account IDs are not sensitive.

Comment (philschmid) [reply infrastructure/ami/hcl2-files/variables.pkr.hcl:37]: > While account IDs, like any identifying information, should be used and shared carefully, they are not considered secret, sensitive, or confidential information. 

https://docs.aws.amazon.com/accounts/latest/reference/manage-acct-identifiers.html#:~:text=While%20account%20IDs%2C%20like%20any,Canonical%20user%20ID

File: infrastructure/ami/scripts/install-huggingface-libraries.sh (hunk 31 above)
Comment (philschmid) [reply infrastructure/ami/scripts/install-huggingface-libraries.sh:12]: Would this install the extras? i don't think thats possible.

Comment (shub-kris) [reply infrastructure/ami/scripts/install-huggingface-libraries.sh:12]: I don't think it will require extras, https://github.com/huggingface/optimum-neuron/blob/main/setup.py#L16

Comment (philschmid) [reply infrastructure/ami/scripts/install-huggingface-libraries.sh:12]: We require the transformers extras with `sklearn,sentencepiece,vision` to make sure training for all model works correctly, e.g. models needing sentence piece or pillow.

**Comment by dacorvo** [comment]
The CI is failing: is this expected ?

**Comment by shub-kris** [comment]
> The CI is failing: is this expected ?

@dacorvo fails because of not having the aws account credentials yet in the repo.

File: .github/workflows/build-ami.yml (hunk 1 above)
Comment (shub-kris) [reply .github/workflows/build-ami.yml]: @glegendre01

File: .github/workflows/build-ami.yml [hunk 32]
```diff
@@ -56,13 +56,13 @@ jobs:
           
           - name: Packer Validate
             id: validate
-            run: packer validate -var "optimum_version=${{ github.event.inputs.tag }}" -var "region=${{ env.AWS_REGION }}" hcl2-files 
+            run: packer validate -var "optimum_version=${{ github.event.inputs.tag || github.event.repository.default_branch }}" -var "region=${{ env.AWS_REGION }}" hcl2-files
```
Comment (philschmid) [inline .github/workflows/build-ami.yml:59]: Why is it here `github.event.repository.default_branch` and not `github.ref`? Wouldn't this mean that when we open a PR with changes, we would still install the main branch? 

Lets align this. Otherwise i can be confusing for others when the need to makes changes for the ami and the main branch.

**Review by philschmid — COMMENTED** [review]
Lets align the ref and then we are good.

File: .github/workflows/build-ami.yml (hunk 32 above)
Comment (shub-kris) [reply .github/workflows/build-ami.yml:59]: Yes, it would then install the main branch. 

The workflow either gets triggered when someone makes changes to `infrastructure/ami/**'`  then in that case the `optimum_version` would be `main` and if it is triggered automatically using scheduler or manually through cli then it would either be `main` or the `tag` passed. 

If we want to build this image at any particular change in the current repo, which makes sense then I need to remove the path: `infrastructure/ami/**`

Comment (philschmid) [reply .github/workflows/build-ami.yml:59]: okay let's keep it but add a comment that others understand why.

Comment (shub-kris) [reply .github/workflows/build-ami.yml]: @philschmid Added the comments

Comment (shub-kris) [reply .github/workflows/build-ami.yml]: So shall I go ahead and merge it?

Comment (philschmid) [reply .github/workflows/build-ami.yml]: Yes feel free to merge
